src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -580,67 +531,65 @@ sealed private[immutable] trait LazyListFactory[+CC[+X] <: LinearSeq[X] with Laz
   * @define Coll `LazyList`
   */
 @SerialVersionUID(3L)
-object LazyList extends LazyListFactory[LazyList] {
-
-  protected def newCons[T](hd: => T, tl: => LazyList[T]): LazyList[T] = new LazyList.Cons(hd, tl)
-
-  @SerialVersionUID(3L)
-  object Empty extends LazyList[Nothing] {
-    override def isEmpty: Boolean = true
-    override def head: Nothing = throw new NoSuchElementException("head of empty lazy list")
-    override def tail: LazyList[Nothing] = throw new UnsupportedOperationException("tail of empty lazy list")
-    def force: this.type = this
-    override def knownSize: Int = 0
-    override def iterator: Iterator[Nothing] = Iterator.empty
-    protected def tailDefined: Boolean = false
-    protected def headDefined: Boolean = false
+object LazyList extends SeqFactory[LazyList] {
+  // Eagerly evaluate cached empty instance
+  private[this] val _empty = newLL(State.Empty).force
+
+  private sealed trait State[+A] extends Serializable {
+    def head: A
+    def tail: LazyList[A]
+    def headDefined: Boolean
   }
 
-  @SerialVersionUID(3L)
-  final class Cons[A](hd: => A, tl: => LazyList[A]) extends LazyList[A] {
-    private[this] var hdEvaluated: Boolean = false
-    private[this] var tlEvaluated: Boolean = false
-    override def isEmpty: Boolean = false
-    override lazy val head: A = {
-      hdEvaluated = true
-      hd
+  private object State {
+    @SerialVersionUID(3L)
+    object Empty extends State[Nothing] {
+      def head: Nothing = throw new NoSuchElementException("head of empty lazy list")
+      def tail: LazyList[Nothing] = throw new UnsupportedOperationException("tail of empty lazy list")
+      def headDefined: Boolean = false
     }
-    override lazy val tail: LazyList[A] = {
-      tlEvaluated = true
-      tl
-    }
-    def force: this.type = {
-      // Use standard 2x 1x iterator trick for cycle detection ("those" is slow one)
-      var these, those: LazyList[A] = this
-      if (!these.isEmpty) {
-        these.head
-        these = these.tail
-      }
-      while (those ne these) {
-        if (these.isEmpty) return this
-        these.head
-        these = these.tail
-        if (these.isEmpty) return this
-        these.head
-        these = these.tail
-        if (these eq those) return this
-        those = those.tail
+
+    @SerialVersionUID(3L)
+    final class Cons[A](private[this] var hd: () => A, val tail: LazyList[A]) extends State[A] {
+      @volatile private[this] var hdEvaluated: Boolean = false
+
+      lazy val head: A = {
+        hdEvaluated = true
+        val res = hd()
+        hd = null // allow GC
+        res
       }
-      this
+
+      def headDefined: Boolean = hdEvaluated
     }
+  }
+
+  private class LazyIterator[+A](private[this] var lazyList: LazyList[A]) extends AbstractIterator[A] {
```

Comment:
Are we certain this should be private?  We seem to rely on it for implementing a fair bit of functionality.  How are library users supposed to replicate the functionality if this is private?  If they can, why don't we?

Reply:
Since it's specific to `LazyList`, it's not clear to me why you'd need to replicate the functionality - you can just call `.iterator`

Reply:
To be clear, I don't feel strongly about it either way